Add wavetable oscillator

brbrr opened the pull request:
Adds a bandlimited wavetable oscillator. The implementation is quite opinionated, and it would require some amount of refactoring to make it general enough to be used in multiple use cases. 

It behaves similarly to `Oscillator` class, with the caveat that one needs to run `Tables::Generate();` before processing the oscillator.

brbrr committed:
Add wavetable oscillator

brbrr commented on Source/Utility/wavetables.h:
```diff
@@ -0,0 +1,390 @@
+    std::vector<WaveBuffer *> buffers;
+};
+
+struct Tables
```
As suggested by @antisvin in the forum, it might be a good idea to provide a templated FFT implementation instead of one hardcoded below.

stephenhensley commented:
ouu, very cool! Thanks for the contribution @brbrr 

I should have some time to give this a look next week!

GuilhermeMoreno committed:
Fix circular dependency in wavetables.cpp

Contributed-by: Guilherme Moreno <guilhermewr9@gmail.com>
Signed-off-by: Guilherme Moreno <guilhermewr9@gmail.com>

GuilhermeMoreno committed:
Add header guards for WaveTable and WaveTableOsc

Contributed-by: Guilherme Moreno <guilhermewr9@gmail.com>
Signed-off-by: Guilherme Moreno <guilhermewr9@gmail.com>

GuilhermeMoreno committed:
Add WaveTables to CMake file

Contributed-by: Guilherme Moreno <guilhermewr9@gmail.com>
Signed-off-by: Guilherme Moreno <guilhermewr9@gmail.com>

GuilhermeMoreno committed:
Converts Tables struct into a template class

Converts the struct to a template class, allowing to provide the FFT implementation as a template parameter.

Signed-off-by: Guilherme Moreno <guilhermewr9@gmail.com>

beserge committed:
Some tweaks to building and interface

Added to .cpp to the makefile, added the .h to daisysp.h,
fixed an include.

Also moved the waveform enum to pass by the type rather than uint8_t. This is a change I want to make across the library to how we handle enums.

beserge committed:
Fix wavetable style

brbrr committed:
Merge branch 'add/wavetable-osc' into fix/cmake-build

brbrr committed:
Merge pull request #2 from GuilhermeMoreno/fix/cmake-build

Fix CMake build

brbrr committed:
Merge pull request #3 from GuilhermeMoreno/feature/wavetable-osc

Converts Tables struct into a template class

beserge committed:
Add wavetables to CMake

beserge committed:
Merge branch 'add/wavetable-osc' of https://github.com/brbrr/DaisySP into add/wavetable-osc

beserge committed:
Fix style once more

beserge commented on Source/Synthesis/wavetable_osc.h:
```diff
@@ -0,0 +1,141 @@
+    {
+        switch(wf)
+        {
+            case Waveform::WAVE_SIN: SetWaveTable(&Tables::Sine); break;
+            case Waveform::WAVE_TRI: SetWaveTable(&Tables::Tri); break;
+            case Waveform::WAVE_SAW: SetWaveTable(&Tables::Saw); break;
+            case Waveform::WAVE_SQUARE: SetWaveTable(&Tables::Square); break;
+            default: SetWaveTable(&Tables::Sine);
```
These now throw the error(s)
`'template<class T, void (* fft_func)(int, T*, T*)> class daisysp::Tables' used without template arguments` and so on.

beserge commented:
Sorry about working on top of you a bit, I think my repo was a bit out of sync and I re-fixed issues that had already been addressed. If we can get it building again with the template change, that would be great!

SimonKunz commented:
Hi all! Any chance this will get merged? Having a wavetable OSC would be so awesome!

Sam2R commented:
Hi All!
Is the wave-table osc has been merged to the library? it's an important step to leverage performance!

davidlpower commented:
Hey, I was wondering why there dosen't seem to be much love from the company on these PRs. I'm curious, is the project dead or being replaced by something (PiSound)?

benjaminvdb commented:
> Hey, I was wondering why there dosen't seem to be much love from the company on these PRs. I'm curious, is the project dead or being replaced by something (PiSound)?

Good question. I'm also wondering about this. Going through the open issues and PRs while contemplating whether to invest learning the platform was a little... off-putting. I really hope it's a temporary glitch :smile:

stephenhensley commented:
The project's not dead, or being replaced.
We're just a small team, and quite busy 😬 

The time I have had for maintaining the libraries has been prioritized on libDaisy lately.
That said, this PR looks like it is pretty much good to go. So I may be able to carve out some time to resolve the build-file conflicts, and merge it soon.
